### ISSUE TITLE
Cleanup FixedPrecisionTensor

### ISSUE THREAD

**Issue by iamtrask** [OpenMined/PySyft#2897]
**Describe the bug**
FixedPrecision tensor has become a catch-all for a myraid of functionality which have nothing to do with fixing precision. At present it seems to include:

0) The ability to model Fixed Precision Numbers using underlying Integer representations
1) BigInt functionality allowing for large precision
2) Convolutional functionality
3) many references to specific subclass types (such as AdditiveSharingTensor).
4) an implementation of pow() which seems unrelated to FixedPrecisionRepresentation
5) Newton-shulz inversion
6) approximation of exponentiation
7) approximation of sigmoid
8) approximation of log
9) etc...

All functionality other than (0) should be removed from this class and refactored into appropriate other classes.

**Comment by iamtrask** [comment]
I will begin working on a properly decoupled BigInt tensor

**Comment by LaRiffle** [comment]
That's right!
1. So actually once you are in Fixed-precision, you can't do anymore pow, exp, simoid, log, inversion the way you did previously, so you have to redefine them, and I'd say this is the good place.
2. Less important, some functions like Convolutional should indeed be put in our dedicated `nn` module where rnn already lies.
3. Having references to AdditiveSharingTensor is *super* annoying, but from our design seems hard to avoid, since we want to operate tensors with different types (FPT & AST)

Actually, we could follow the design principle of another lib and define fixed-precision not as a tensor type but as an encoder which is attached to a tensor. This sounds weird but:
- It reconciliate FPT & AST in a single type and solves 3.
- It diminishes the number of tensors types and reduces the tensor chain size
- It is more readable for non-expert users

To avoid the limitation of having to redefine the encoder for the HomomorphicTensor, etc, encoder could be supplied by a inheritance mechanism, but then @karlhigley is not going to be happy lol

In a nutshell
- I don't think 1. is that of a problem
- Issue 2. can be solved easily
- Issue 3. needs more intelligence, to make for example the mul function of FPT readable

**Comment by karlhigley** [comment]
My only ask would be to supply the encoder as a constructor param with a default value (instead of as an additional parent class via inheritance.) 🙂

**Comment by iamtrask** [comment]
I don't see the semantic difference between an encoder and a tensor type
(something used in our tensor chain).

**Comment by karlhigley** [comment]
I'm not sure I fully understand it either, just chiming in on the point about inheritance. 😄

**Comment by LaRiffle** [comment]
The difference here is that using the encoder, you don't add the FPT node in the tensor chain, so you're simpler & faster. But this reflexion is far from being top priority for the moment^^

**Comment by ADMoreau** [comment]
I'll move the conv2d. Should I put it in the nn module or should I put it in some other class. If another class, are there any suggestions? I jsut want to make sure I'm sticking to a design strategy I may not be aware of.

**Comment by LaRiffle** [comment]
In nn is good!
But maybe conv2d has to transformed into a class, or maybe not, I let you choose the best option here :)

**Comment by iamtrask** [comment]
Important question conv2d get _exactly_ the same values in the same
positions as torch's Conv2d? Aka, can I train with PyTorch's conv2d then
swap it out for this one and have it automatically work?

We've been working on an implementation which is this precise in nn,  but
it doesn't have all features yet.

**Comment by ADMoreau** [comment]
Sure I'll give it a shot

**Comment by stephenjfox** [comment]
@ADMoreau Are you still working on this?
@iamtrask is this still relevant?

**Comment by ADMoreau** [comment]
No, I'm not working on this.

**Comment by sachin-101** [comment]
@LaRiffle @iamtrask  I would like to work on this. Two things I will try for a start are:
1. Comment out unnecessary functionalities in FixedPrecisionTensor. And make sure the rest of the code works.
2. Move conv2d implementation from FixedPrecisionTensor to torch.nn

**Comment by sachin-101** [comment]
@LaRiffle Based on this [tutorial on Build your own tensor](https://github.com/OpenMined/PySyft/blob/master/examples/tutorials/advanced/Build%20your%20own%20tensor%20type%20(advanced).ipynb) and your comment above 

> So actually once you are in Fixed-precision, you can't do anymore pow, exp, simoid, log, inversion the way you did previously, so you have to redefine them, and I'd say this is the good place.

I would let the above-mentioned functions be in precision.py
And all functions which are not specific to how we operate with FixedPrecisionTensor would be commented out/moved.
